File: app/src/main/java/io/github/hrilab/pepper_realtime/ChatMessage.java
package io.github.hrilab.pepper_realtime;

public class ChatMessage {
    public enum Sender {
        USER, ROBOT
    }
    
    public enum Type {
        REGULAR_MESSAGE,
        FUNCTION_CALL,
        IMAGE_MESSAGE
    }

    private String message;
    private final Sender sender;
    private final Type type;
    private String imagePath; // optional
    private String itemId; // optional, for tracking Realtime API conversation items
    
    // Function call specific fields
    private String functionName;
    private String functionArgs;
    private String functionResult;
    private boolean isExpanded = false;

    // Private constructor 
    private ChatMessage(Sender sender, Type type) {
        this.sender = sender;
        this.type = type;
    }

    // Static factory for function call messages
    
    public static ChatMessage createFunctionCall(String functionName, String functionArgs, Sender sender) {
        ChatMessage chatMessage = new ChatMessage(sender, Type.FUNCTION_CALL);
        chatMessage.functionName = functionName;
        chatMessage.functionArgs = functionArgs;
        chatMessage.message = ""; // Will be generated in display
        return chatMessage;
    }

    // Keep original constructor for backward compatibility
    public ChatMessage(String message, Sender sender) {
        this(sender, Type.REGULAR_MESSAGE);
        this.message = message;
    }

    // Keep original image constructor for backward compatibility
    public ChatMessage(String message, String imagePath, Sender sender) {
        this(sender, Type.IMAGE_MESSAGE);
        this.message = message;
        this.imagePath = imagePath;
    }

    // Getters and setters
    public String getMessage() { return message; }
    public void setMessage(String message) { this.message = message; }

    public Sender getSender() { return sender; }
    public Type getType() { return type; }

    public String getImagePath() { return imagePath; }
    
    public String getItemId() { return itemId; }
    public void setItemId(String itemId) { this.itemId = itemId; }
    
    public String getFunctionName() { return functionName; }
    public String getFunctionArgs() { return functionArgs; }
    public String getFunctionResult() { return functionResult; }
    public void setFunctionResult(String functionResult) { this.functionResult = functionResult; }
    
    public boolean isExpanded() { return isExpanded; }
    public void setExpanded(boolean expanded) { this.isExpanded = expanded; }
}

File: app/src/main/java/io/github/anonymous/pepper_realtime/ui/ChatMessage.kt
package io.github.anonymous.pepper_realtime.ui

import java.util.UUID

/**
 * Represents a chat message in the conversation UI.
 * Supports regular messages, image messages, and function call messages.
 */
class ChatMessage private constructor(
    val sender: Sender,
    val type: Type,
    val uuid: String
) {
    enum class Sender {
        USER, ROBOT
    }

    enum class Type {
        REGULAR_MESSAGE,
        FUNCTION_CALL,
        IMAGE_MESSAGE
    }

    var message: String = ""
    var imagePath: String? = null
    var itemId: String? = null
    var functionName: String? = null
    var functionArgs: String? = null
    var functionResult: String? = null
    var isExpanded: Boolean = false

    /**
     * Creates a shallow copy of the message with updated text.
     * Preserves the UUID to maintain item identity for DiffUtil.
     */
    fun copyWithNewText(newText: String): ChatMessage {
        return ChatMessage(sender, type, uuid).also {
            it.message = newText
            it.imagePath = this.imagePath
            it.itemId = this.itemId
            it.functionName = this.functionName
            it.functionArgs = this.functionArgs
            it.functionResult = this.functionResult
            it.isExpanded = this.isExpanded
        }
    }

    /**
     * Creates a shallow copy of the message with updated function result.
     * Preserves the UUID to maintain item identity for DiffUtil.
     */
    fun copyWithFunctionResult(result: String): ChatMessage {
        return ChatMessage(sender, type, uuid).also {
            it.message = this.message
            it.imagePath = this.imagePath
            it.itemId = this.itemId
            it.functionName = this.functionName
            it.functionArgs = this.functionArgs
            it.functionResult = result
            it.isExpanded = this.isExpanded
        }
    }

    companion object {
        /**
         * Static factory for function call messages
         */
        @JvmStatic
        fun createFunctionCall(functionName: String, functionArgs: String, sender: Sender): ChatMessage {
            return ChatMessage(sender, Type.FUNCTION_CALL, UUID.randomUUID().toString()).also {
                it.functionName = functionName
                it.functionArgs = functionArgs
                it.message = "" // Will be generated in display
            }
        }
    }

    // Secondary constructors for backward compatibility

    /**
     * Constructor for regular text messages
     */
    constructor(message: String, sender: Sender) : this(sender, Type.REGULAR_MESSAGE, UUID.randomUUID().toString()) {
        this.message = message
    }

    /**
     * Constructor for image messages
     */
    constructor(message: String, imagePath: String?, sender: Sender) : this(sender, Type.IMAGE_MESSAGE, UUID.randomUUID().toString()) {
        this.message = message
        this.imagePath = imagePath
    }
}

